src/boot/chid.h

```diff
@@ -45,20 +55,38 @@ assert_cc(offsetof(Device, descriptor) == 0);
 assert_cc(offsetof(Device, chid) == 4);
 assert_cc(offsetof(Device, devicetree.name_offset) == 20);
 assert_cc(offsetof(Device, devicetree.compatible_offset) == 24);
+assert_cc(offsetof(Device, fw.name_offset) == 20);
+assert_cc(offsetof(Device, fw.compatible_offset) == 24);
 assert_cc(sizeof(Device) == 28);
 
 static inline const char* device_get_name(const void *base, const Device *device) {
-        if (device->descriptor != DEVICE_DESCRIPTOR_DEVICETREE)
+        uint32_t off;
+        switch (DEVICE_TYPE_FROM_DESCRIPTOR(device->descriptor))
+        {
+        case DEVICE_TYPE_DEVICETREE:
+                off = device->devicetree.name_offset;
+                return off == 0 ? NULL : (const char *) ((const uint8_t *) base + off);
+        case DEVICE_TYPE_FW:
+                off = device->fw.name_offset;
+                return off == 0 ? NULL : (const char *) ((const uint8_t *) base + off);
```

Comment:
Lines 68 and 71 are common between all the types, so move to the end of this function.